[RFC] ST sensor drivers migration to STdC

avisconti opened the pull request:
STdC is a standard package written by ST that aims to provide a standard register i/f for ST MEMS
sensors. It is already offered/used in other s/w systems as well.
The idea behind it is to have a clean and stable functions/defines to read/write sensors registers.
Official public repo is: 
https://github.com/STMicroelectronics/STMems_Standard_C_drivers

The license seems to be the same as stm32cube, so there should be no issues here.
Since STdC package is including math.h I had to enable NEWLIB_LIBC.

In this RFC I did an experimental test using lis2dw12 driver. It works for both i2c and spi.
Please provide your feedback.

avisconti committed:
ext/hal: Introduce STMEMS Standard C Drivers

This package contains platform independent drivers written in C
language for STMicroelectronics sensors. The aim of this package
is to provide a common, clean and stable interface to access sensor
registers.

Library is located in ext/hal/st/stmemsc/

Origin: ST Microelectronics
License: BSD-3-Clause
URL: https://www.st.com/en/embedded-software/c-driver-mems.html
Commit: v1.00
Purpose: provide a common and stable i/f to access sensor registers
Maintained-by: ST Microelectronics

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
@erwango 
This is not urgent as I'm planning it for v1.15, but would be great to receive your feedback
for the 'ext/' part as soon as possible. If it makes sense to proceed with this standard sensor reg i/f 
drivers I have some porting work to do.

avisconti commented:
@MaureenHelm 
I would like to seek your opinion as well.

avisconti commented:
@erwango @MaureenHelm 
I re-aligned to latest commit and re-pushed.
As I was saying this RFC is quite important for me, as if accepted I would need to make modifications
to #13062 and to #13478 outstanding PR.

erwango reviewed (changes requested):
Few comments on packages integration, otherwise looks good.

avisconti commented on ext/hal/st/stmemsc/README:
```diff
@@ -0,0 +1,42 @@
+   BSD-3-Clause
+
+License Link:
+   http://www.st.com/resource/en/license_agreement/dm00218346.pdf
```
We have the same agreement cited in all stm32cube packages, like ext/hal/st/stm32cube/stm32f4xx/README for example.

Why should be an issue here?

erwango replied:
Right, let me sort this out.

erwango replied:
@avisconti , I'm updating the link for Cube packages: https://github.com/zephyrproject-rtos/zephyr/pull/15731

avisconti replied:
> @avisconti , I'm updating the link for Cube packages: #15731

OK, I'm going to change it as well. Thx!

avisconti commented on ext/hal/st/stmemsc/README.md:
```diff
@@ -0,0 +1,111 @@
+1 - Introduction
```
Correct! I removed all the examples. So I will strip off from the README.md all references to examples.

avisconti commented on ext/hal/st/stmemsc/a3g4250d_STdC/README.txt:
```diff
@@ -0,0 +1,17 @@
+
```
Mmmmh ok.
So I see two options now. Either keeping also the examples in the package or adapting all the
README (but there are many) stripping of all references to examples.

@erwango 
If you don't see issues I may go with option one which seems to be less impacting.

erwango replied:
I'm not sure adding examples juts because README files refer to it is a great option.
Do we really need these README files at all?
I don't see much added value in the information provided, this is only copy/paste from one to the other, and most of the information is not valid as referring to examples that you initially considered as superfluous (and I think you were right).
The only interesting information that differ from one file to the other is `H3LIS331DL DS rev3.0 `, which could be factorized in a upper file if needed.

avisconti replied:
So, your preferred option would be:
1. Package w/o examples
2. Remove all README files under drivers
3. Keep the top README file w/o reference to example

Correct?
I'm ok with it. The only thing is that everytime I have to re-align the package there is an
extra effort (but not that much) to clean the package up.

erwango replied:
This is indeed indeed my preferred way to do it.
For the extra pain to do it, I suggest to spend few hours on a script to do it for  you.
I have this for cube packages update and this saves me painful hours each time.

avisconti commented:
rebased and repushed as agreed with @erwango

avisconti commented:
@erwango 
It seems that you are in charge of everything under ext/hal/st.
This will include also stmemsc.

avisconti commented:
I got now that I have to solve a couple of false postives in UNDEF_KCONFIG_WHITELIST.
:-(

erwango commented:
> @erwango
It seems that you are in charge of everything under ext/hal/st.
This will include also stmemsc.

@avisconti, as discussed, can you add your name in front of stmemsc in CODEOWNER?

avisconti commented:
> @avisconti, as discussed, can you add your name in front of stmemsc in CODEOWNER?

Yes I did.
It is in 52bc149 commit. Is it not matching your expectation? If I remember well we concluded
to have the 'backup' line (i.e. /ext/hal/st) first and then the line for stmemsc, correct?

Now I made a change in the ci-tools whitelist in order to have the CI passing for this PR.

erwango commented:
> It is in 52bc149 commit. Is it not matching your expectation?

It's perfect, much more than my monday eyes.

avisconti committed:
driver/sensor: lis2dw12: make use of STdC definitions

Port the lis2dw12 sensor driver on top of the lis2dw12_StdC
HAL interface (under ext/hal/st/stmemsc/).

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
recheck

avisconti commented:
@erwango  @galak 
I see that now the CI is giving an issue when running tests. The issue is that to properly compile we
need to define following macro:

CONFIG_NEWLIB_LIBC=y
CONFIG_NEWLIB_LIBC_FLOAT_PRINTF=y

Before I used 'select' directly in stmemsc Kconfig directly, but I saw that usage of select for visible
symbols (the ones with a prompt) is not recommended. So, I recently changed it in this PR avoiding
'select' and using instead 'depends on'. But this requires the symbols to be defined somewhere during
CI test. I have now two options:

1. Going back to using 'select'
2. Defining somewhere the two symbols. Maybe in sensors_i_z.conf (and all other sensors conf files)
    but this would impact other sensors too. 

So, can you help me in using the right approach for fixing CI?

galak commented:
> @erwango @galak
> I see that now the CI is giving an issue when running tests. The issue is that to properly compile we
> need to define following macros:
> 
> CONFIG_NEWLIB_LIBC=y
> CONFIG_NEWLIB_LIBC_FLOAT_PRINTF=y
> 
> Before I used 'select' directly in stmemsc Kconfig file, but I saw that usage of 'select' for visible
> symbols (the ones with a prompt) is not recommended. So, I recently changed it in this PR avoiding
> 'select' and using instead 'depends on'. But this requires the symbols to be defined somewhere during
> CI test. I have now two options:
> 
> 1. Going back to using 'select'
> 2. Defining somewhere the two symbols. Maybe in sensors_i_z.conf (and all other sensors conf files)
>    but this would impact other sensors too.
> 
> So, can you help me in using the right approach for fixing CI?
> 
> EDIT:
> Maybe also @MaureenHelm may add some comments.
> 
> EDIT2:
> It seems that there is a third options discussed with @MaureenHelm, which is to add a new test
> for sensor using the new stmemsc external hal, where I can define a separate conf file.

So I think you should have a separate test for this if newlib is required for the sensor drivers.  We shouldn't burden sensors_i_z.conf with that requirement.

If that makes its easier to use 'depends' than great.  However, I'd try 'imply' and see how that works.

avisconti commented:
re-pushed again
Now I'm using 'imply' inside the stmemsc Kconfig.
Moreover I defined two new sensor tests for LIS2DW12: test_build_sensors_stmemsc and test_build_sensors_stmemsc_trigger.
All ST sensors that will make use of stmemsc hal in the future will be put inside this test.

avisconti commented:
@erwango @MaureenHelm 
Not sure what the Codeowners failed check is about, the details are not clear enough.

I think the code now is in a good shape, isn't it?

erwango commented:
@avisconti , seems there was an issue in CODEOWNER file that prevented correct parsing.
It is fixed now. Can you rebase to take fix into account and see if fixing you issue?

avisconti commented:
> @avisconti , seems there was an issue in CODEOWNER file that prevented correct parsing.
> It is fixed now. Can you rebase to take fix into account and see if fixing you issue?

OK, thx!
re-pushed! Let's see how it goes now.

avisconti commented:
@MaureenHelm 
What's ur opinion about introducing the STdC HAL i/f?
If you approve I can later go further porting also LPS22HH and LSM6DSO on it.
Thx!

avisconti commented:
@galak @MaureenHelm 
Anything else recommended in this PR?

avisconti commented:
@erwango 
Can you confirm again the license (BSD-3-Clause) is not an issue?
I think it is the same as STM32cube, right?

erwango commented:
@avisconti , BSD-3 Clause is fine, sure. 
Though, there is a indeed a shade vs Apache which is too subtle for the License bot to handle. This is hence up to the maintainer to ack.

erwango commented:
@nashif, do you want to merge this one now? or latter once the modules are there?

avisconti commented:
@nashif @erwango 
I rebased this PR on the latest code.
Moreover I changed the ext/hal/st/stmemsc/Kconfig file with two simple changes:
- I added the USE symbols for all drivers (before was only lis2dw12)
- I removed the prompt from the USE symbols, as we want to force them from the KConfig of each specific driver.

avisconti commented:
@nashif @erwango 
I moved the STdC in hal_st module. I'm going to close this PR.
The lis2dw12 porting on top of STdC HAL i/f will be resubmitted as a new PR later on.
